transformer_engine/pytorch/csrc/thd_utils.cuh

```diff
@@ -183,59 +183,199 @@ __global__ void thd_lse_kernel(lse_dtype *lse, float *half_lse, int *cu_seqlens,
 }
 
 /***************************************************************************************************
- * Support THD format for Context Parallel: Out correction in forward
+ * Support BSHD, SBHD, and THD formats for Context Parallel: Out correction in forward
  **************************************************************************************************/
 
-template <typename dtype, int only_second_half, int tile_size, bool lse_packed>
-__global__ void thd_out_correction_kernel(dtype *out, dtype *out_per_step, float *lse,
-                                          float *lse_per_step, int *cu_seqlens, int batch,
-                                          int num_heads, int dim_per_head, int lse_seqlen,
-                                          int lse_per_step_seqlen) {
+// format of out and lse, ignoring d as it’s always the last dimension.
+enum QKVFormat { SBH, BSH, BHS, HBS, TH, HT };
+
+template <int n>
+struct TensorList {
+  void *addresses_out[n];
+  void *addresses_lse[n];
+  int start_tensor_this_launch;
+};
+
+// describe tensor format for simplified computation.
+template <QKVFormat format>
+struct TensorFormat {
+  // store the bsht order for simplified computation, where bsht corresponds to 0, 1, 2, 3, and store_format[3] marks whether bs is fused into t
+  // For example, for the SBH format, the values of store_format are {1, 0, 2, 0}; for the TH format, the values of store_format are {3, 2, any-value, 1}
+  int8_t store_format[4];
+  int *cu_seqlens_s;
+  // size of tensor, b s h t
+  int size[4];
```

Comment:
I think size array here means the size of each dimension? If so, why don't you have some variables like `batch_size`, `seq_length`, `num_heads`, `head_dim`, etc? the variable naming is more clear and easier to understand, right?